modules/tensor_mechanics/src/actions/GeneralizedPlaneStrainAction.C

```diff
@@ -0,0 +1,105 @@
+    Action(params),
+    _displacements(getParam<std::vector<NonlinearVariableName> >("displacements")),
+    _ndisp(_displacements.size()),
+    _scalar_strain_zz(getParam<NonlinearVariableName>("scalar_strain_zz"))
+{
+  if (_ndisp!= 2)
+    mooseError("GeneralizedPlaneStrain only works for two dimensional case!");
+}
+
+void
+GeneralizedPlaneStrainAction::act()
+{
+  if (_current_task == "add_kernel")
+  {
+    std::string k_type("GeneralizedPlaneStrainOffDiag");
+    InputParameters params = _factory.getValidParams(k_type);
+
+    params.set<std::vector<NonlinearVariableName> >("displacements") = _displacements;
+
+    if (isParamValid("temperature"))
+      params.set<NonlinearVariableName>("temperature") = getParam<NonlinearVariableName>("temperature");
+
+    params.set<std::vector<VariableName> >("scalar_strain_zz") = {_scalar_strain_zz};
+
+    if (isParamValid("base_name"))
+      params.set<std::string>("base_name") = getParam<std::string>("base_name");
+
+    if (isParamValid("block"))
+      params.set<std::vector<SubdomainName> >("block") = getParam<std::vector<SubdomainName> >("block");
+
+    params.set<bool>("use_displaced_mesh") = getParam<bool>("use_displaced_mesh");
+
+    for (unsigned int i = 0; i < _ndisp; ++i)
+    {
+      std::string k_name = "GeneralizedPlaneStrainOffDiag_" + _name + "_" + Moose::stringify(i);
+      params.set<NonlinearVariableName>("variable") = _displacements[i];
+
+      _problem->addKernel(k_type, k_name, params);
+    }
+  }
+  else if (_current_task == "add_user_object")
+  {
+    std::string uo_type("GeneralizedPlaneStrainUserObject");
+    InputParameters params = _factory.getValidParams(uo_type);
+
+    std::string uo_name = "GeneralizedPlaneStrainUserObject_" + _name;
+
+    if (isParamValid("base_name"))
+      params.set<std::string>("base_name") = getParam<std::string>("base_name");
+
+    params.set<FunctionName>("traction_zz") = getParam<FunctionName>("traction_zz");
+    params.set<Real>("factor") = getParam<Real>("factor");
+
+    if (isParamValid("block"))
+      params.set<std::vector<SubdomainName> >("block") = getParam<std::vector<SubdomainName> >("block");
+
+    params.set<MultiMooseEnum>("execute_on") = "linear";
+    params.set<bool>("use_displaced_mesh") = getParam<bool>("use_displaced_mesh");
+
+    _problem->addUserObject(uo_type, uo_name, params);
+  }
+  else if (_current_task == "add_scalar_kernel")
+  {
+    std::string sk_type("GeneralizedPlaneStrain");
+    InputParameters params = _factory.getValidParams(sk_type);
+
+    std::string sk_name = "GeneralizedPlaneStrain_" + _name;
+
+    params.set<NonlinearVariableName>("variable") = _scalar_strain_zz;
+
+    // set the UserObjectName from previously added UserObject
+    params.set<UserObjectName>("generalized_plane_strain") = "GeneralizedPlaneStrainUserObject_" + _name;
+
+    _problem->addScalarKernel(sk_type, sk_name, params);
+  }
+}
```

Comment:
This is OK, but in general you can replace nearly all of this code by using `InputParameters::applyParameters()`, we need to get better at trying to use that method.